source/jcli/cli.d

```diff
@@ -2,327 +2,1675 @@ module jcli.cli;
+            .matchAndExecute(tokenizer, errorHandler);
     }
 
-    private
+    /// Resolves the invoked command by parsing the arguments array,
+    /// executes the `onExecute` method of any intermediary command groups,
+    /// Returns the result of the execution, and whether there were errors.
+    MatchAndExecuteContext matchAndExecute(scope string[] args)
+    {
+        auto tokenizer = argTokenizer(args);
+        return _matchAndExecute(tokenizer);
+    }
+    
+    /// ditto
+    /// Uses the default error handler. 
+    MatchAndExecuteContext matchAndExecute(Tokenizer : ArgTokenizer!T, T)
+    (
+        scope ref Tokenizer tokenizer
+    )
     {
-        Resolver!CommandInfo _resolver;
-        CommandInfo[] _uniqueCommands;
-        CommandInfo _default;
-        string _appName;
+        auto handler = DefaultParseErrorHandler();
+        return _matchAndExecute(tokenizer, handler);
     }
+}
 
-    this()
+enum SpecialThings
+{
+    none = 0,
+    help = 1
+}
+
+SpecialThings tryMatchSpecialThings(Tokenizer : ArgTokenizer!TRange, TRange)
+(
+    ref scope Tokenizer tokenizer
+)
+{
+    alias Kind = ArgToken.Kind;
+    const currentToken = tokenizer.front;
+
+    outerSwitch: switch (currentToken.kind)
+    {
+        case Kind.fullNamedArgumentName:
+        case Kind.shortNamedArgumentName:
+        {
+            switch (currentToken.nameSlice)
+            {
+                case "help":
+                case "h":
+                {
+                    tokenizer.popFront();
+                    
+                    // TODO: handle this better
+                    // if (tokenizer.front.kind == Kind.namedArgumentValue)
+                    // {
+                    //     writeln("Please, do not give `help` a value.");
+                    //     tokenizer.popFront();
+                    // }
+
+                    return SpecialThings.help;
+                }
+                default:
+                    break outerSwitch;
+            }
+        }
+
+        case Kind.orphanArgument:
+        case Kind.positionalArgument:
+        {
+            switch (currentToken.nameSlice)
+            {
+                case "/?":
+                case "/help":
+                case "/h":
+                {
+                    tokenizer.popFront();
+                    return SpecialThings.help;
+                }
+                default:
+                    break outerSwitch;
+            }
+        }
+
+        default:
+            break;
+    }
+
+    return SpecialThings.none;
+}
+
+struct MatchAndExecuteContext
+{
+    State _state;
+
+    union
+    {
+        SpecialThings _specialThing;
+        ExecuteCommandResult _executeCommandResult;
+        string _matchedName;
+        string _notMatchedName;
+    }
+
+    // TODO: maybe store them inline
+    // TODO: calculate the max path length in the graph and set that as max size.
+    struct StorageItem
+    {
+        int typeIndex;
+        void* commandPointer;
```

Comment:
It's stored internally this way, in fact. We could and should expose it to the user via a type-safe wrapper. Using it directly like this internally is totally ok imo.

Reply:
The index can be mapped back to the type by just indexing Types. The reason I store it as an index is because the graph works with indices+it's context independent, essentally

Reply:
What I mean by that, is that the context struct is just data. Whatever is on top of it decides what to do with in on its own.